lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -418,6 +453,56 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
     return success();
   }
 
+  LogicalResult handlePublicModuleRefPorts(FModuleOp module) {
+    // Decls must be at the top mlir module scope
+    auto declBuilder = ImplicitLocOpBuilder::atBlockBegin(
+        module.getLoc(), getOperation().getBody());
+    auto builder = ImplicitLocOpBuilder::atBlockBegin(module.getLoc(),
+                                                      circuitOp.getBodyBlock());
+
+    for (size_t portIndex = 0, numPorts = module.getNumPorts();
+         portIndex != numPorts; ++portIndex) {
+      auto refType = module.getPortType(portIndex).dyn_cast<RefType>();
+      if (!refType || isZeroWidth(refType.getType()) ||
+          module.getPortDirection(portIndex) != Direction::Out)
+        continue;
+      auto portValue =
+          module.getArgument(portIndex).cast<mlir::TypedValue<RefType>>();
+
+      Attribute ref;
+      SmallString<128> stringLeaf;
+      if (failed(resolveReferencePath(portValue, builder, ref, stringLeaf)))
+        return failure();
+
+      SmallString<128> formatString;
+      if (ref)
+        formatString += "{{0}}";
+      formatString += stringLeaf;
+
+      // Insert a macro with the format:
+      // ref_<circuit name>_<module name>_<ref name> <internal path
+      // from module>
+      auto macroName = builder.getStringAttr("ref_" + circuitOp.getName() +
+                                             "_" + module.getName() + "_" +
+                                             module.getPortName(portIndex));
```

Comment:
What would happen if the circuit/main module, module, or ref are later renamed? Should this also be renamed or should it not? This is more of a question for the ABI doc about something like:

```
circuit always:
  module always:
    wire module: UInt<1>
```

Is this: `ref_always_always_module module_0`, `ref_always_0_always_0_module_0 module_0`, or something else?

I think it's reasonable to make a case that this follows the original FIRRTL name even if the ref changes. I.e., I think the current code is correct. The spec should probably make this explicit.